### DISCUSSION TITLE
an actual async `AsyncIterator` example

### DISCUSSION THREAD

**Discussion by folkertdev** [PyO3/pyo3#4234] [Questions]
hi, does someone have a practical example of a rust stream/async iterator that is exposed to python as an async iterator?

this page https://pyo3.rs/main/migration#iteranextoutput-are-deprecated has an example, but what it implements is not actually an async iterator?! there are no awaits on the rust side anyway, so its asynchronous behavior is trivial. I want something like this

```rust
    async fn __anext__(&mut self) -> PyResult<Option<tsp::ReceivedTspMessage>> {
        self.0.next().await.transpose().map_err(py_exception)
    }
```

but apparently the `__anext__` method cannot be an `async fn`, even if I make it simpler

```rust
    async fn __anext__(&mut self) -> usize {
        0
    }
```

I get this error (no idea what to do with that)

```
36  | #[pymethods]
    | ^^^^^^^^^^^^ the trait `IntoPyCallbackOutput<_>` is not implemented for `impl futures::Future<Output = usize>`
    |
    = help: the following other types implement trait `IntoPyCallbackOutput<Target>`:
              <bool as IntoPyCallbackOutput<bool>>
              <bool as IntoPyCallbackOutput<i32>>
              <usize as IntoPyCallbackOutput<isize>>
              <usize as IntoPyCallbackOutput<usize>>
              ... some stuff omitted
            and 3 others
note: required by a bound in `AsyncIterBaseTag::convert`
   --> /home/folkertdev/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.21.2/src/impl_/pymethods.rs:402:16
    |
400 |     pub fn convert<Value, Target>(self, py: Python<'_>, value: Value) -> PyResult<Target>
    |            ------- required by a bound in this associated function
401 |     where
402 |         Value: IntoPyCallbackOutput<Target>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `AsyncIterBaseTag::convert`
```

this post https://github.com/PyO3/pyo3/discussions/4064 seems related but never got a proper answer.

**Comment by jregistr** [comment]
Heys, I implemented an example async iterator for streaming API responses. Perhaps this could be helpful? https://github.com/jregistr/python-stream-async-pyo3/blob/main/stream-q/src/lib.rs#L142
